### PR TITLE
Create a tutorial

### PR DESCRIPTION
**Pull request by t0mhaines**
+ [x] Setup insomnia
+ [x] Create a request
+ [x] Templating introduction
+ [x] Post request
+ [x] Using Tags
+ [x] Authentication Overview

### CODE REVIEW

**Comment by t0mhaines** [comment]
@nelsonic are you able to give this a quick once over for typos etc before merging?

**Comment by nelsonic** [comment]
@th0mas will read it through today after standup. 👀

File: README.md [hunk 1]
```diff
@@ -1,2 +1,256 @@
...
+https://insomnia.rest/download/
+
+Make sure you download and install **Insomnia Core**
+
+##### Example Server
+
+For this tutorial, we will be using an example rest server. Clone this repository,
+
+```
+git clone https://github.com/dwyl/learn-nerves
```
Comment (nelsonic) [inline README.md]: should this be `learn-nerves` or `learn-insomnia` ?

File: README.md [hunk 2]
```diff
@@ -1,2 +1,256 @@
...
+```
+
+In `learn-nerves` you should find a folder called `server`, and within that
```
Comment (nelsonic) [inline README.md]: No `server` directory in `learn-nerves`: https://github.com/dwyl/learn-nerves
image

File: README.md (hunk 1 above)
Comment (t0mhaines) [reply README.md]: This should be `learn-insomnia` oops

File: README.md (hunk 2 above)
Comment (t0mhaines) [reply README.md]: There should be in this repo, sorry for the typo

**Review by nelsonic — CHANGES_REQUESTED** [review]
Hi @th0mas, 
This is a _good_ effort towards explaining **`insomnia`**, 👍 
_however_ the "How?" section covers a **`Go`** Todo List Api. 🤷‍♂️ 
We don't use **`Go`** for anything and we don't _plan_ to use it.
We use `Elixir` and we need our **`insomnia`** guide to be _relevant_ to the language we use.

I can only apoligise if I didn't make it clear in the issue yesterday: https://github.com/dwyl/learn-insomnia/issues/1
But we _really_ need the `insomnia` tutorial to be _relevant_ to what we are building. i.e. use `Elixir/Phoenix`: 
use-elixir-for-insomnia
If we are using `insomnia` for our https://github.com/dwyl/smart-home-auth-server 
then our _tutorial_ needs to help the person who is new to the project _understand_ why/how insomnia is used within that context. 

Our established [Technology Stack](https://github.com/dwyl/technology-stack) is `Elixir/Phoenix`. If it's not _possible_ to use `Elixir` with `insomnia`, then please make that clear in the `README.md`. 

The _last_ thing we want to do is introduce people to a brand new ***programming language*** while they are learning **insomnia** it adds too much complexity to the tutorial and ultimately isn't relevant to the context where we are _using_ insomnia in "production". It actually ends up being more _confusing_ than helpful. 

> It might be safe to assume that people at your **Uni course** know **`Go-lang`**, it's even reasonably safe to assume that _I_ (Nelson) know it, but it's _definitely_ not a safe assumption when it comes to the _rest_ of the team/community. 😞 


Just to re-iterate: in ***isolation*** an **`insomnia`** tutorial that features a **`new`** (_to @dwyl_) programming language is perfectly fine and you've done a **_good_ job** of making it simple and approachable. But in ***context***, where we are using **`insomnia`** with **`Elixir`** for the Smart Home Hub, it's not very _useful_ for whoever wants to _understand_ how to _maintain_ the project(s) we are building. 

Please don't _scrap_ the `Go` work you've done, it might be relevant to the devs who use `Go`.
But please move it to a new folder called `go-lang-example` along with the instructions.

What we _need_ from this tutorial is more:
"_This is a **super basic example** showing how we are using **`insomnia` in our `Elixir` project**(s)_"

**Comment by t0mhaines** [comment]
Insomnia is a HTTP API client that only interacts with remote servers (and by extension, programming languages) over HTTP.

For the purpose of the guide, I assumed no programming knowledge whatsoever, as its not needed to consume APIs through Insomnia, and Insomnia is purely a HTTP **client**.

I used Golang as it means we can ship a server to test against as a single executable for people that may not have programming knowledge. This means it's only one command (`./server`) to start a test API server, with no prior dependencies required. 

 If I had used *Elixir* - (which I still can), users would have had to download and install an Elixir toolchain, along with Phoenix and possibly PostgreSQL, for no benefit in learning how to use Insomnia.

I can also include a Phoenix Server, but this would not change anything about the tutorial itself, as the REST API would stay the same, the same way that writing a server in Go or Elixir does not change what web browser you use to access the server.

Hope this helps

**Comment by nelsonic** [comment]
Hi @th0mas, totally get that **`insomnia`** is an HTTP API Client. 👍 
In this case it's interacting with a _local_ server over HTTP. 
And if the person wants to understand the server "under the hood", they need to read `Go`.
`Go` is very readable and I think we should _keep_ the server you have created.
But it's definitely a "context switch" for someone who is not used to reading it.
I don't think it would "_tax_" anyone in our _current_ team, it's not exactly complex.
But I've seen people who are learning get stuck and unable to make progress over smaller barriers.
So I'm always conscious of minimising the barriers wherever possible.
Maybe I shouldn't care as much about people getting stuck ...
But as a self-taught coder (_for the most part alone_) I know how frustrating it can be.
Anyway, the point isn't the `Go` code you have written, which is readable, it's the _principal_ of reducing obstacles to learning wherever possible. And contextualising things in terms of where we are _using_ it. If someone wants to _understand_ the HTTP server and they already know `Elixir`, it's going to be _much_ faster if the HTTP server is in the language they know. 💭 

What I was saying in the issue was that they _already_ have `Elixir` installed. 
Assume that the person reading the `learn-insomnia` is _maintaining_ the code you have written for the firmware, they are in the `Elixir` mindset already and just need to grok what this "insomnia" thing is so they can get back to maintaining the firmware. 

100% agree we definitely don't need a full-on Phoenix Server with Postgres. 👍 
Luckily a plug-based web App/API can be written very simply in `Elixir`.
e.g. https://github.com/dwyl/elixir-plug-tutorial/blob/master/lib/app/router.ex
just in-line the router handlers in the `router.ex` file and you're off to the races. 
Not saying `Elixir` is _better_ than `Go` by any stretch.
But when I see a Binary executable being added to a Git Repo, I shudder:

image

It's generally undesirable to store blobs of compiled code in Git. 
Even if in _this_ case it actually makes _sense_ from a usability perspective;
having the `/server/server` _executable_ makes it a "black box".
It's just not a _precedent_ you want to set because beginners will think it's "OK" to do it.

Completely understand that the tutorial itself won't change with an `Elixir` web server.
In fact that's _exactly_ what we want! We _want_ the tutorial to be language "agnostic".
But if the person wants to _understand_ the code in the server, we don't want them to have to go _googling_ for things. 

Thanks.

**Comment by t0mhaines** [comment]
@nelsonic Just pushed some changes to add a test admin route on Elixir, might need to `git pull` on your end to finish the tutorial 👍

**Review by nelsonic — APPROVED** [review]
@th0mas thanks for adding the Elixir Server.
100% _agree_ that for this simple Example the `Go` is _much_ friendlier for people who don't have much experience of Elixir (_or other languages_). 👍 
I've added a few extra screenshots while following the tutorial. (_and fixed a couple of typos_)
But your original content is solid. 
Thanks. 🙌

**Pull request merged**